MGMT-15704  Assisted service should create Day2 import CR for hub cluster.

paul-maidment opened the pull request:
When using assisted-service in KubeAPI mode, we want to import the local cluster automatically, allowing it to be managed in a similar fashion to the spoke cluster.

The intent is to allow the user to perform day2 operations on the local cluster, to allow the addition of workers and so on.

Presently this is only possible via manual efforts and is not very customer friendly.

This PR aims to resolve this by adding functionality to import the local cluster as described above.

The behaviour during upgrade will be dealt with in another ticket.

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

You can refer to [Kubernetes community documentation] on writing good commit messages, which provides good tips and ideas.

Some PRs address specific issues. Please, refer to the [CONTRIBUTING] documentation for more
information on how to link a PR to an existing issue.

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer?
- Is this PR relying on CI for an e2e test run?
- Should this PR be tested in a specific environment?
- Any logs, screenshots, etc that can help with the review process?

-->

## List all the issues related to this PR

- [ ] New Feature <!-- new functionality -->
- [ ] Enhancement <!-- refactor, code changes, improvement, that won't add new features -->
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD <!-- Notice that changes for Dockerfiles/Jenkinsfiles aren't tested in CI due to a known bug. -->

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

<!-- Please, select one or more if needed: -->

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

filanov commented:
/assign @nmagnezi

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+		i.log.Errorf("hubClusterDay2Importer: There was an error fetching the AgentClusterInstall while checking for need to register local cluster.")
+		return false, err
+	}
+	if agentClusterInstall != nil {
```
please add a todo here with the relevant ticket to handle upgrade

paul-maidment replied:
Done

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+	if !shouldImportLocalCluster {
+		message := "hubClusterDay2Importer: no need to import local cluster as registration already detected"
+		i.log.Info(message)
+		return errors.New(message)
```
no need to return an error in this case, nothing have failed

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+	}
+	openshift_version := ""
+	release_image := ""
+	if clusterVersion != nil && clusterVersion.Status.History[0].State == configv1.CompletedUpdate {
```
is it always the first?

paul-maidment replied:
Seems that way, yes, the State is the crucial thing, it will be `configv1.PartialUpdate` until the installation has completed.
But yes, I tested this assumption using `oc adm upgrade` -- the first history item is always the latest

filanov replied:
can you install older ocp and upgrade to make sure it will remain the first one?

paul-maidment replied:
I will try this now to verify

paul-maidment replied:
I can confirm that the latest version is always the [0]th entry

`        "history": [
            {
                "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=True (), so the update from 4.13.9 to 4.13.11 is probably neither recommended nor supported.",
                "completionTime": null,
                "image": "quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64",
                "startedTime": "2023-09-06T08:52:43Z",
                "state": "Partial",
                "verified": false,
                "version": "4.13.11"
            },
            {
                "completionTime": "2023-09-06T08:52:09Z",
                "image": "quay.io/openshift-release-dev/ocp-release@sha256:a266d3d65c433b460cdef7ab5d6531580f5391adbe85d9c475208a56452e4c0b",
                "startedTime": "2023-09-06T08:19:17Z",
                "state": "Completed",
                "verified": false,
                "version": "4.13.9"
            }
        ],
`

paul-maidment replied:
Confirmed, latest is always the [0]th item in the history...

```
[
  {
    "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.14.0-ec.4-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"ClusterVersionUpgrade
able\" failed because of \"AdminAckRequired\": Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions.\n* Precondition \"EtcdRecentBackup\" failed because of \"UpgradeBackupInProgress\": RecentBackup: Backup pod phase: \"P
ending\"\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.11 to 4.14.0-ec.4 is unknown.",
    "completionTime": null,
    "image": "quay.io/openshift-release-dev/ocp-release:4.14.0-ec.4-x86_64",
    "startedTime": "2023-09-06T12:37:19Z",
    "state": "Partial",
    "verified": false,
    "version": "4.14.0-ec.4"
  },
  {
    "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=True (), so
 the update from 4.13.9 to 4.13.11 is probably neither recommended nor supported.",
    "completionTime": "2023-09-06T10:06:05Z",
    "image": "quay.io/openshift-release-dev/ocp-release:4.13.11-x86_64",
    "startedTime": "2023-09-06T08:52:43Z",
    "state": "Completed",
    "verified": false,
    "version": "4.13.11"
  },
  {
    "completionTime": "2023-09-06T08:52:09Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:a266d3d65c433b460cdef7ab5d6531580f5391adbe85d9c475208a56452e4c0b",
    "startedTime": "2023-09-06T08:19:17Z",
    "state": "Completed",
    "verified": false,
    "version": "4.13.9"
  }
]

```

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+		openshift_version = clusterVersion.Status.History[0].Version
+		release_image = clusterVersion.Status.History[0].Image
+		i.log.Infof("hubClusterDay2Importer: Found openshift_version %s", openshift_version)
+	} else {
```
suggesting to change the order of the if
```go
if clusterVersion == nil || clusterVersion.Status.History[0].State != configv1.CompletedUpdate {
		message := "hubClusterDay2Importer: Cluster version info is empty, cannot proceed"
		i.log.Error(message)
		return errors.New(message)
}
```
then you will not need the else and the code is clearer that way.

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+		i.log.Errorf("hubClusterDay2Importer: Unable to fetch local cluster kubeconfigs due to error %s", err.Error())
+		return err
+	}
+	if kubeConfigSecret != nil {
```
is there a case where you have no error and the kubeconfig is nil?

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+		return errors.New(message)
+	}
+
+	dns, err := i.clusterImportOperations.GetDNS("cluster")
```
it's a bit strange to have function that gets a const

paul-maidment replied:
This is easier to mock, the underlying operation can get any DNS entry. I'd rather keep this as is

filanov replied:
maybe make it GetClusterDNS it will be clearer and you will not need to pass this parameter

filanov reviewed (commented):
Please add in the commit message why did you had to add permissions to specific resources

General comment about if else 
usually it's cleaner to handle the error first, then you will not need the else in most cases

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,240 @@
+			Platform: hivev1.Platform{
+				AgentBareMetal: &agent.BareMetalPlatform{
+					AgentSelector: metav1.LabelSelector{
+						MatchLabels: map[string]string{"infraenv": fmt.Sprintf("%s-day2-infraenv", LocalClusterNamespace)},
```
you can remove this part, you don't need to decide for the user what will be the infra env. 
you can write something like local cluster or something generic

nmagnezi commented on cmd/main.go:
```diff
@@ -205,6 +206,19 @@ func maxDuration(dur time.Duration, durations ...time.Duration) time.Duration {
+	err := localClusterImport.ImportLocalCluster()
+	if err != nil {
+		// Failure to import the local cluster is not fatal but we should warn in the log.
+		log.Warnf("Could not import local cluster into ACM due to error %s", err.Error())
```
```suggestion
		log.Warnf("Could not import local cluster due to error %s", err.Error())
```

nmagnezi replied:
I don't it should mention ACM specifically:
1. It might not be relevant to some deployment types we support.
2. ACM is a downstream name.

paul-maidment replied:
What is the correct terminology for what we are doing here?

paul-maidment replied:
After discussion with @filanov we have settled on "import local cluster" as the terminology to be used.

filanov replied:
+1 to what Nir suggested, we have both ACM and MCE, so his suggestion cover both

nmagnezi commented on cmd/main.go:
```diff
@@ -641,6 +655,10 @@ func main() {
 	}()
 
+	if Options.EnableKubeAPI {
+		importLocalCluster(ctrlMgr, log)
```
Is there no feature flag for this import, specifically?

filanov replied:
+1 we probably need to have something that can disable this feature if needed

paul-maidment replied:
adding a flag to allow feature to be disabled.

nmagnezi commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,226 @@
+		return false, err
+	}
+	if agentClusterInstall != nil {
+		i.log.Infof("hubClusterDay2Importer: Found AgentClusterInstall for hub cluster, assuming that hub has been correctly registered for ZTP day 2 operations: %s", agentClusterInstall.Name)
```
Are there any checks we can do here to ensure that cluster is actually healthy?
Also, why is ZTP mentioned here?

paul-maidment replied:
I'm not sure that this is possible at this stage.

The one thing we do perform is a check of the cluster-version history, which will show up as completed if the last update was applied.

filanov replied:
i'm not sure which tests can be done in this case, domain resolution and valid kubeconfig, but if you are taking them from the cluster and even if they are invalid, what can be done?

nmagnezi commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,226 @@
+	// The presence of agentClusterInstall for the local cluster will determine whether or not we need to perform an import
+	agentClusterInstall, err := i.clusterImportOperations.GetAgentClusterInstall(LocalClusterNamespace, LocalClusterNamespace+"-cluster-install")
+	if err != nil && !k8serrors.IsNotFound(err) {
+		i.log.Errorf("hubClusterDay2Importer: There was an error fetching the AgentClusterInstall while checking for need to register local cluster.")
```
Why do all the errors have a `hubClusterDay2Importer` prefix?

paul-maidment replied:
I did this to make things more "greppable" but if it's non standard or weird, I can remove it?
I think that if you are trying to debug a specific piece of functionality, having some text to grep helps with filtering
that's the main reason I did this

paul-maidment replied:
Removing the prefix

nmagnezi commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,226 @@
+	if !shouldImportLocalCluster {
+		message := "hubClusterDay2Importer: no need to import local cluster as registration already detected"
+		i.log.Info(message)
+		return errors.New(message)
```
+1  , I was about to comment the same thing.
In line 37 you treated this as an Info log.

paul-maidment replied:
already fixed

nmagnezi commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,226 @@
+		i.log.Errorf("hubClusterDay2Importer: Unable to fetch local cluster kubeconfigs due to error %s", err.Error())
+		return err
+	}
+	i.log.Infof("hubClusterDay2Importer: Found secret %s", kubeConfigSecret.Name)
```
1. This should probably be a debug.
2. Please specify the details of the related cluster and/or also the secret namespace - to make it more specific.

nmagnezi commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,228 @@
+	return true, nil
+}
+
+func (i *LocalClusterImport) ImportLocalCluster() error {
```
Few things about this function:
1. It does to many things, which should be split to sub functions for both readability and tests.
2. It can potentially fail in too many places (which is related to the first point, in part). I also think that for many things here it is worth to use multiErr to aggregate a much of the problems and reflect them to the user as early as possible.

paul-maidment replied:
Adding a feature flag to options

paul-maidment replied:
1: Done
2: Done

paul-maidment commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,226 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const LocalClusterNamespace = "local-cluster"
```
I'm not 100% sure that this is in line with the goals of the feature. The main goal I see is that we want to simplify the process of management of the hub cluster by creating some of the CRs the user needs. 

Maybe we could accept an env var optionally and default to "local-cluster"

Also note that "local-cluster" is the name of the cluster that is created in multi cluster hub.

filanov replied:
the only option to set env vars at the moment is with the unsupported configmap.
i don't think that it will be wise to add it as an official api at the moment, lets get a feedback on that in the next release and if needed we  will add it as an option

filanov replied:
but env var with a default value can be a good option even if it is set via unsupported configmap

paul-maidment commented:
> Using ZTP in the commit message is incorrect, here you are referring to kube-api or the operator, have nothing to do with ZTP

Fixed now

> Using ZTP in the commit message is incorrect, here you are referring to kube-api or the operator, have nothing to do with ZTP

Fixed

filanov commented on cmd/main.go:
```diff
@@ -205,6 +208,23 @@ func maxDuration(dur time.Duration, durations ...time.Duration) time.Duration {
 
+func importLocalCluster(ctrlMgr manager.Manager, log *logrus.Logger) {
+	if !Options.EnableLocalClusterImport {
+		log.Debug("EnableLocalClusterImport disabled in options, skipping...")
```
can be info, it's not something periodic

filanov commented on cmd/main.go:
```diff
@@ -162,6 +163,8 @@ var Options struct {
 	PreprovisioningImageControllerConfig controllers.PreprovisioningImageControllerConfig
 	BMACConfig                           controllers.BMACConfig
+	EnableLocalClusterImport             bool   `envconfig:"ENABLE_LOCAL_CLUSTER_IMPORT" default:"true"`
+	LocalClusterImportNamespace          string `envconfig:"LOCAL_CLUSTER_IMPORT_NAMESPACE" defualt:"local-cluster"`
```
please make sure with ACM/MCE  that they are not using the same namespace

filanov commented on pkg/localclusterimport/local_cluster_import_operations.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (o *LocalClusterImportOperations) CreateClusterImageSet(clusterImageSet *hivev1.ClusterImageSet) (*hivev1.ClusterImageSet, error) {
+	err := o.cachedApiClient.Create(o.context, clusterImageSet)
```
what will happen is it's already exists? 
same question to all the resources.

paul-maidment replied:
It would fail to write the cluster image set and the old one would remain in place. 
Not sure what the best thing to do is, on one hand, we could delete anything we find when performing this write, but seems to me that this would not be the smart choice until we have looked into the impact of the update ticket.

filanov replied:
until we handle upgrade it should be good enough

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,288 @@
+
+func (i *LocalClusterImport) ImportLocalCluster() error {
+
+	shouldImportLocalCluster, err := i.shouldImportLocalCluster()
```
As we talked, this function can be removed and you will try to create all the objects and if they are already exist you will ignore them, using `k8serrors.IsAlreadyExists(err)` to check the error

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,276 @@
+	}
+	// Fetch the recently stored AgentClusterInstall so that we can obtain the UID
+	aci, err := i.clusterImportOperations.GetAgentClusterInstall(i.localClusterNamespace, i.localClusterNamespace+"-cluster-install")
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
```
get cannot return already exist

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	err = i.createNamespace(i.localClusterNamespace)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
```
minor comment - you already handling already exist error inside `createNamespace`

paul-maidment replied:
I actually changed this, I need to be able to mock this so I moved the handling of this stuff into import_local_cluster

filanov commented on pkg/localclusterimport/import_local_cluster.go:
```diff
@@ -0,0 +1,276 @@
+
+	if kubeConfigSecret != nil {
+		err = i.createAdminKubeConfig(kubeConfigSecret)
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
```
minor comment - you already handling already exist error inside `createAdminKubeConfig`

paul-maidment replied:
I actually changed this, I need to be able to mock this so I moved the handling of this stuff into import_local_cluster

paul-maidment commented:
/retest

paul-maidment commented:
/cherry-pick release-ocm-2.8
